man/ammt.8
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "AMMT" 8 "" "" ""
.SH NAME
ammt \- AMANDA version of mt
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 5
\fBammt\fR [\-d] [\fB\-f\fR | \fB\-t\fR | \fB\fIdevice\fR\fR] \fIcommand\fR [\fIcount\fR]
.ad
.hy

.SH "DESCRIPTION"

.PP
\fBAmmt\fR provides just enough of the standard UNIX \fBmt\fR command for the needs of \fBAMANDA\fR\&. This is handy when doing a full restore and the standard \fBmt\fR program has not yet been found\&.

.PP
\fBAmmt\fR also provides access to the \fBAMANDA\fR output drivers that support various tape simulations\&.

.PP
See the \fBamanda\fR(8) man page for more details about \fBAMANDA\fR\&. See the \fBOUTPUT DRIVERS\fR section of \fBamanda(8)\fR for more information on the \fBAMANDA\fR output drivers\&.

.SH "OPTIONS"

.TP
\fB\-d\fR
Turn on debugging output\&.

.TP
\fB\-f\fR\fI device\fR
Access tape device \fBdevice\fR\&. If not specified, the \fBTAPE\fR environment variable is used\&.

.TP
\fB\-t\fR\fI device\fR
Same as \fB\-f\fR\&.

.TP
\fBcommand\fR\fB count\fR
Which command to issue, and an optional count of operations\&.

.SH "COMMANDS"

.PP
Each command may be abbreviated to whatever length makes it unique\&.

.TP
\fBeof|weof\fR\fB count\fR
Write \fBcount\fR (default: 1) end of file marks (tapemarks)\&.

.TP
\fBfsf\fR\fB count\fR
Skip forward \fBcount\fR (default: 1) files\&.

.TP
\fBbsf\fR\fB count\fR
Skip backward \fBcount\fR (default: 1) files\&.

.TP
\fBasf\fR\fB count\fR
Position to file number \fBcount\fR (default: 0) where zero is beginning of tape\&. This is the same as a \fBrewind\fR followed by a \fBfsf\fR  \fBcount\fR\&.

.TP
\fBrewind\fR
Rewind to beginning of tape\&.

.TP
\fBoffline|rewoffl\fR
Rewind to beginning of tape and unload the tape from the drive\&.

.TP
\fBstatus\fR
Report status information about the drive\&. Which data reported, and what it means, depends on the underlying operating system, and may include:

.RS

.TP
ONLINE
Indicates the drive is online and ready\&.

.TP
OFFLINE
Indicates the drive is offline or not ready\&.

.TP
BOT
Indicates the drive is at beginning of tape\&.

.TP
EOT
Indicates the drive is at end of tape\&.

.TP
PROTECTED
Indicates the tape is write protected\&.

.TP
ds
Device status\&.

.TP
er
Error register\&.

.TP
fileno
Current tape file number\&.

.TP
blkno
Current tape block number file\&.

.RE
.IP

.RS
.Sh "Note"
Many systems only report good data when a tape is in the drive and ready\&.

.RE

.PP


.SH "AUTHOR"

.PP
Marc Mengel <mengel@fnal\&.gov>, John R\&. Jackson <jrj@purdue\&.edu>: Original text

.PP
Stefan G\&. Weichinger, <sgw@amanda\&.org>, maintainer of the \fBAMANDA\fR\-documentation: XML\-conversion

.SH "SEE ALSO"

.PP
\fBamanda\fR(8)



NEWS
Changes in release 2.5.1b2

* debug file are seperated in subdir client/server/amandad with the config name.
* amandates client config option.
* gnutar_list-dir client config option.
* works on solaris 10 64bits.

Changes in release 2.5.1b1

* Klocwork defects fixed.
* Coverity defect fixed.
* work with tar-1.15.91
* Open SSL encryption support
* amfetchdump -o is replaced by -O.
* amrecover_do_fsf default to yes
* amrecover_check_label default to yes
* amcheck -w will do all test including the tape writable test.
  use amcheck -t -w if you want only the tape writable test.
* usetimestamps
* bsdudp/bsdtcp, 2 new auth.
* Unlimited number of DLE on a client with bsdtcp/rsh or ssh auth.
* docs/howto-auth.txt
* amandad_path, client_username and ssh_keys (for ssh/rsh)
* client config file (see man amanda-client.conf)
  Only use by amrecover.
* amrecover use the security-api
* amoldrecover which use the old amindexd/amidxtaped protocol
* -o option to overwrite config option
  (see in man amanda, the CONFIGURATION OVERWRITE section)
* new format of .amandahosts
* new format of xinetd.d/amanda

Changes in release 2.5.0p2

* listhost subcommand in amrecover

Changes in release 2.5.0p1

* Add the 'amtape update' command.

Changes in release 2.5.0

* Communication security/authentication: Kerberos 4/5, OpenSSH
* Data security: Symmetric/Assymetric encrytion algorithms (aesutil
  and gpg encryption),  Encryption can be done on server or client,
  Custom encryption utilities can be used.
* Compression: Ability to add custom compression tools. This is a
  really useful feature especially since it can specified for a DLE.  You
  can use different compression algorithm for images, binaries, ascii files
  and so on.
* Dump images spanning multiple media volumes - Dump images are no
  longer restricted to a single media volume (tape or vtape).  Data restoration
  can be done using amrecover and amfetchdump commands.
* Auto tape labelling - This optional feature is good for disk backups.

Change in release 2.4.5p1

* New [--days <num>] argument to 'amadmin <conf> tape'.
* New amoverview -last -num0 -togo0 options.

Change in release 2.4.5

* new displayunit global option to select the unit use to display number
  k=kilo, m=mega, g=giga, t=tera.
* new amoverview -skipmissed option.

Change in release 2.4.5b1

* holding disk disk use timestamped directory.
* autoflush flush today's dump.
* new bumppercent global option, this should improve bumping criteria,
  the bumpsize is set to a fixed value which can be adequate for small
  and large disk.
* bumpsize, bumppercent, bumpdays and bumpmult can be in a dumptype.
* calcsize support include and exclude like gnutar.
* new 'estimate' dumptype option to select estimate type:
    CLIENT: estimate by the dumping program.
    CALCSIZE: estimate by the calcsize program, a lot faster but less acurate.
    SERVER: estimate based on statistic from previous run, take second but
            can be wrong on the estimate size.

Changes in release 2.4.4p3

* new chg-iomega changer script.
* amanda will not use a tape if it's label is not in the tapelist file
  (eg. after an amrmtape).

Changes in release 2.4.4p2

* new initial_poll_delay config in chg-zd-mtx
* 'amadmin find' list disk removed from the disklist.
* amrecover can recover a disk removed from the disklist file.
* amanda works with samba3.
* new chg-disk changer script to use with the file: driver.
* Kerberos 4 support integrated and functional again.

Changes in release 2.4.4p1

* amverify do not advance to next tape at the end.
* new amflush -b and -s options, change -f behavior.
* new chg-mcutil changer.
* amrecover_changer works with chg-multi.
* default driveslot for chg-zd-mtx is now 0.
* amplot generate color postscript by default.
* amplot -b flag to generate b/w postscript.
* rait: works with file:
* print a postscript label for each tape used.

Changes in release 2.4.4

* New -a option to amcheck to always send an email.

Changes in release 2.4.4b1

* maxpromoteday: New option for a dumptype.
* New amtapetype program (replace tapetype)
* Client compile on CYGWIN
* amrestore -f <fileno> : amrestore can do fsf before reading the tape
* amrestore -l <label>  : amrestore can check the label before restoring.
* New config options:
  amrecover_do_fsf: Amrecover will use the -f flag of amrestore to position the
                    tape.
  amrecover_check_label: amrecover will pass the -l flag to amrestore.
  amrecover_changer: amrecover will use your changer.
  maxdumpsize: The maximum size of the dumps during a run,
               default to tapesize * runtapes
  taperalgo: select your algo for the taper, try 'largestfit', it can improve
             tape usage.
* amrecover allow to retry a skip a tape.
* New --days option to 'amadmin <config> balance' command.
* New --date option to amstatus
* amreport print an usage by tape.
* Newer promote algorithm that try to reduce the number of full on a
  single host in a run.
* New changer: chg-juke, chg-null and chg-rait

Changes in release 2.4.3

* Few small bug fixed
* New runtapes argument to amverify.
* New amverifyrun program.
* 2.4.3 client works with older server.

Changes in release 2.4.3b4

* new noop packet type. It allow the server to know the client features.
* exclude list with relative path is working again.
* amstatus is working again.

Changes in release 2.4.3b3

* --with-maxtapeblocksize configure options
* blocksize tapetype option
* file-pad tapetype option
* Multiple exclude in dumptype
* Option include in dumptype
* New disklist syntax:
*   hostname diskname [ diskdevice ] dumptype [ spindle [ interface ] ]
* chg-zd-mtx: Major cleanup and general overhaul.
* amrecover: new listdisk command.

Changes in release 2.4.3b2

* amrestore try to find the next chunk in the current directory.
* amrecover: You can change the tape device between each tape.
* amstatus work with amflush.

Changes in release 2.4.3b1

* tapeio.

tape-src/ammt.c
#ifdef NO_AMANDA
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/stat.h>

#include "output-rait.h"

extern char *getenv();

#define	tape_open	rait_open
#define	tapefd_read	rait_read
#define	tapefd_write	rait_write
#define	tapefd_fsf	rait_tapefd_fsf
#define	tapefd_rewind	rait_tapefd_rewind
#define	tapefd_status	rait_tapefd_status
#define	tapefd_unload	rait_tapefd_unload
#define	tapefd_weof	rait_tapefd_weof
#define tapefd_setinfo_length(outfd, length)
#define	tapefd_close	rait_close

#else
#include "amanda.h"
#include "tapeio.h"
#endif

static int do_asf();
static int do_bsf();
static int do_status();

struct cmd {
    char *name;
    int min_chars;
    int count;
    int (*func)();
    int flags;
} cmd[] = {
    { "eof",		0,	1,	tapefd_weof, O_RDWR },
    { "weof",		0,	1,	tapefd_weof, O_RDWR },
    { "fsf",		0,	1,	tapefd_fsf, O_RDONLY },
    { "asf",		0,	0,	do_asf, O_RDONLY },
    { "bsf",		0,	1,	do_bsf, O_RDONLY },
    { "rewind",		0,	0,	tapefd_rewind, O_RDONLY },
    { "offline",	0,	0,	tapefd_unload, O_RDONLY },
    { "rewoffl",	0,	0,	tapefd_unload, O_RDONLY },
    { "status",		0,	0,	do_status, O_RDONLY },
    { NULL,		0,	0,	NULL }
};

static char *pgm;
static int debug_ammt = 0;

static char *tapename;

static int
do_asf(fd, count)
    int fd;
    int count;
{
    int r;

    if(debug_ammt) {
	fprintf(stderr, "calling tapefd_rewind()\n");
    }
    if(0 != (r = tapefd_rewind(fd))) {
	return r;
    }
    if(debug_ammt) {
	fprintf(stderr, "calling tapefd_fsf(%d)\n", count);
    }
    return tapefd_fsf(fd, count);
}

static int
do_bsf(fd, count)
    int fd;
    int count;
{
    if(debug_ammt) {
	fprintf(stderr, "calling tapefd_fsf(%d)\n", -count);
    }
    return tapefd_fsf(fd, -count);
}

static int
do_status(fd, count)
    int fd;
    int count;
{
    int ret;
    struct am_mt_status stat;

    if(debug_ammt) {
	fprintf(stderr, "calling tapefd_status()\n");
    }
    if((ret = tapefd_status(fd, &stat)) != 0) {
	return ret;
    }
    printf("%s status:", tapename);
    if(stat.online_valid) {
	if(stat.online) {
	    fputs(" ONLINE", stdout);
	} else {
	    fputs(" OFFLINE", stdout);
	}
    }
    if(stat.bot_valid && stat.bot) {
	fputs(" BOT", stdout);
    }
    if(stat.eot_valid && stat.eot) {
	fputs(" EOT", stdout);
    }
    if(stat.protected_valid && stat.protected) {
	fputs(" PROTECTED", stdout);
    }
    if(stat.device_status_valid) {
	printf(" ds == 0x%0*lx",
	       stat.device_status_size * 2,
	       (unsigned long)stat.device_status);
    }
    if(stat.error_status_valid) {
	printf(" er == 0x%0*lx",
	       stat.error_status_size * 2,
	       (unsigned long)stat.error_status);
    }
    if(stat.fileno_valid) {
	printf(" fileno == %ld", stat.fileno);
    }
    if(stat.blkno_valid) {
	printf(" blkno == %ld", stat.blkno);
    }

    putchar('\n');
    return 0;
}

static void
usage()
{
    fprintf(stderr, "usage: %s [-d] [-f|-t device] command [count]\n", pgm);
    exit(1);
}

int
main(int argc, char **argv) {
    int ch;
    int count;
    int i;
    int j;
    int fd;
    int save_errno;
    char *s;

    if((pgm = strrchr(argv[0], '/')) != NULL) {
	pgm++;
    } else {
	pgm = argv[0];
    }
    tapename = getenv("TAPE");
    while(-1 != (ch = getopt(argc, argv, "df:t:"))) {
	switch(ch) {
	case 'd':
	    debug_ammt = 1;
	    fprintf(stderr, "debug mode!\n");
	    break;
	case 'f':
	case 't':
	    tapename = optarg;
	    break;
	default:
	    usage();
	    /* NOTREACHED */
	}
    }
    if(optind >= argc) {
	usage();
	/* NOTREACHED */
    }

    /*
     * Compute the minimum abbreviation for each command.
     */
    for(i = 0; cmd[i].name; i++) {
	cmd[i].min_chars = 1;
	while (1) {
	    for(j = 0; cmd[j].name; j++) {
		if(i == j) {
		    continue;
		}
		if(0 == strncmp(cmd[i].name, cmd[j].name, cmd[i].min_chars)) {
		    break;
		}
	    }
	    if(0 == cmd[j].name) {
		break;
	    }
	    cmd[i].min_chars++;
	}
	if(debug_ammt) {
	    fprintf(stderr, "syntax: %-20s -> %*.*s\n",
			    cmd[i].name,
			    cmd[i].min_chars,
			    cmd[i].min_chars,
			    cmd[i].name);
	}
    }

    /*
     * Process the command.
     */
    s = "unknown";
    j = strlen(argv[optind]);
    for(i = 0; cmd[i].name; i++) {
	if(0 == strncmp(cmd[i].name, argv[optind], j)) {
	    if(j >= cmd[i].min_chars) {
		break;
	    }
	    s = "ambiguous";
	}
    }
    if(0 == cmd[i].name) {
	fprintf(stderr, "%s: %s command: %s\n", pgm, s, argv[optind]);
	exit(1);
    }
    optind++;
    if(0 == tapename) {
	fprintf(stderr, "%s: -f device or -t device is required\n", pgm);
	exit(1);
    }
    if(debug_ammt) {
	fprintf(stderr, "tapename is \"%s\"\n", tapename);
    }

    count = 1;
    if(optind < argc && cmd[i].count) {
	count = atoi(argv[optind]);
    }

    if(debug_ammt) {
	fprintf(stderr, "calling tape_open(\"%s\",%d)\n", tapename, cmd[i].flags);
    }
    if((fd = tape_open(tapename, cmd[i].flags)) < 0) {
	goto report_error;
    }

    if(debug_ammt) {
	fprintf(stderr, "processing %s(%d)\n", cmd[i].name, count);
    }
    if(0 != (*cmd[i].func)(fd, count)) {
	goto report_error;
    }

    (void)tapefd_close(fd);

    exit(0);

report_error:

    save_errno = errno;
    fprintf(stderr, "%s %s", tapename, cmd[i].name);
    if(cmd[i].count) {
	fprintf(stderr, " %d", count);
    }
    errno = save_errno;
    perror(" failed");
    exit(1);
}
